skel/daemons/license.go

```diff
@@ -0,0 +1,203 @@
+			log.Println(err.Error())
+			continue
+		}
+
+		expireTime := time.Unix(exp, 0)
+		diff := expireTime.Sub(time.Now())
+
+		// Check if license need renewal
+		if len(lic) > 0 && diff.Minutes() > 30 { // FIXME: Adjust time before renewal
+			log.Println("License renewal was skipped")
+			continue
+		}
+
+		err = createNewLicense(app, vaultService)
+		if err != nil {
+			log.Println(err.Error())
+		}
+
+		time.Sleep(time.Minute * 15) // FIXME: Adjust poll period
+	}
+}
+
+func createNewLicense(app *sonic.Runtime, vaultService *vault.Service) error {
+	platformID, err := getPlatformIDFromVault(vaultService)
+	if err != nil {
+		return err
+	}
+
+	opendaxConfig := app.Conf.Opendax
+	url, err := url.Parse(opendaxConfig.Addr)
+	if err != nil {
+		return err
+	}
+	url.Path = path.Join(url.Path, "/api/v2/opx/licenses/new")
+
+	privateKey, err := getPrivateKeyFromVault(vaultService)
+	if err != nil {
+		return err
+	}
+
+	keyStore := jwt.KeyStore{}
+	err = keyStore.LoadPrivateKey(privateKey)
+	if err != nil {
+		return err
+	}
+	t := jwtgo.New(jwtgo.GetSigningMethod("RS256"))
+	jwtToken, err := t.SignedString(keyStore.PrivateKey)
+
+	req, err := http.NewRequest(http.MethodPost, url.String(), nil)
+	if err != nil {
+		return err
+	}
+	// Add request header
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("PlatformID", platformID)
+	req.Header.Add("Authorization", "Bearer "+jwtToken)
+
+	// Call HTTP request
+	httpClient := &http.Client{}
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	// Convert response body to []byte
+	resBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return err
+	}
+
+	// Check for API error
+	if res.StatusCode != http.StatusCreated {
+		return err
+	}
+
+	license := LicenseResponse{}
+	err = json.Unmarshal(resBody, &license)
+	if err != nil {
+		return err
+	}
+
+	err = saveLicenseToVault(vaultService, license.License)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getPlatformIDFromVault(vaultService *vault.Service) (string, error) {
+	app := "peatio"
+	scope := "private"
+	key := "platform_id"
+
+	// Load secret
+	vaultService.LoadSecrets(app, scope)
+
+	// Get secret
+	result, err := vaultService.GetSecret(app, key, scope)
+	if err != nil {
+		return "", err
+	}
+
+	return result.(string), nil
+}
+
+func getPrivateKeyFromVault(vaultService *vault.Service) (string, error) {
+	app := "sonic"
+	scope := "secret"
+	key := "jwt_private_key"
+
+	// Load secret
+	vaultService.LoadSecrets(app, scope)
+
+	// Get secret
+	result, err := vaultService.GetSecret(app, key, scope)
+	if err != nil {
+		return "", err
+	}
+
+	return result.(string), nil
+}
+
+func getLicenseFromVault(vaultService *vault.Service) (string, error) {
+	app := "finex"
+	scope := "secret"
+
+	// Load secret
+	vaultService.LoadSecrets(app, scope)
+
+	// Get secret
+	license, err := vaultService.GetSecret(app, "license", scope)
```

Comment:
Let's use finex_license_key as the field name here as in https://github.com/openware/openfinex/blob/02142e24f2af8555087f669bf79ccab90a996802/pkg/config/config.go#L16

Reply:
Renamed.